### DISCUSSION TITLE
Setting custom headers

### DISCUSSION THREAD

**Discussion by RReverser** [yandeu/five-server#76] [Help / Questions]
Is it possible to set custom headers from the config file? It's especially important when working on WebAssembly projects, see https://github.com/ritwickdey/vscode-live-server/issues/657 for the discussion on the original project.

**Comment by yandeu** [comment]
If you use a javascript config file, you can simply add an array of custom middlewares and add/remove any header you want.

See: https://github.com/yandeu/five-server/blob/d39fac4d2d4ad53873b9c6327e64fd536cf69b7a/src/types.ts#L40

Something like:

```js
(req, res, next) => {
  console.log('Time:', Date.now())
  next()
}
```

**Reply by usirg** [reply]
@yandeu 

I'm interested in setting a custom response header "Content-Encoding: br" to be able to test brotli compressed Unity builds.

I have a fiveserver.config.js which works fine and for testing I added the following:

```typescript
  middleware: [
    (req, res, next) => {
      console.log('Time:', Date.now())
      next()
    }
  ],
``` 

There are no errors but I also don't get any output in the console.

I have little practice with JS/TS, so maybe I am missing something obvious.

Do I understand correctly that this is how to set response headers in the above function:

```typescript
res.setHeader('Content-Encoding','br')
``` 
Any pointers would be greatly appreciated!

**Reply by yandeu** [reply]
If you are using the vscode extension, I believe, it will not show the log.

Maybe try the npm version to see the console.log with `npx five-server`.

**Reply by usirg** [reply]
I gave this another shot and based on [serve-unity](https://github.com/TheRoccoB/serve-unity/blob/master/serve-unity.js), I implemented a middleware function that sets the headers for both Content-Encoding and Content-Type for paths ending in br and gz:

```javascript
// fiveserver.config.js

const path = require("path");

module.exports = {
  watch: 'Builds',
  root: 'Builds/build',
  open: ['index.html'],
  host: '0.0.0.0',
  highlight: true, // enable highlight feature
  injectBody: true, // enable instant update
  remoteLogs: true, // enable remoteLogs
  injectCss: false, // disable injecting css
  navigate: false, // enable auto-navigation
  https: true,
  middleware: [
    (req, res, next) => {
      const mimeType = getMimeType(req.url)
      if (req.url.endsWith('.br')) {
        res.setHeader('Content-Encoding', 'br');
        res.setHeader('Content-Type', mimeType);
      } else if (req.url.endsWith('.gz')) {
        res.setHeader('Content-Encoding', 'gzip');
        res.setHeader('Content-Type', mimeType);
      }
      next();
    }
  ]
};

const MIME_TYPES = {
  ".html": "text/html",
  ".js": "application/javascript",
  ".wasm": "application/wasm",
  ".data": "application/octet-stream",
  ".json": "application/json",
  ".css": "text/css",
};

// Helper function to get the original file MIME type
function getMimeType(uri) {
  if (uri.endsWith(".gz")) {
    return MIME_TYPES[path.extname(uri.replace(".gz", ""))] || "application/octet-stream";
  } else if (uri.endsWith(".br")) {
    return MIME_TYPES[path.extname(uri.replace(".br", ""))] || "application/octet-stream";
  }
  return MIME_TYPES[path.extname(uri)] || "application/octet-stream";
}
``` 

With this, five server successfully serves brotli-compressed Unity WebGL builds😁

**Comment by RReverser** [comment]
Hmm. Sure, that's one option, thanks, but... any chance of adding this to declarative VSCode settings? (so that I could configure it in my editor for all projects & sync settings with other machines where I use VSCode)

**Reply by yandeu** [reply]
Maybe I could add a custom headers array.
What headers are missing and why does it have issues with wasm? I often use wasm myself, and it always works fine.

**Reply by RReverser** [reply]
As mentioned int he linked issue, it's WebAssembly threads specifically that can't work without COOP/COEP headers, so any projects that uses Wasm threads can't work with stock HTTP servers that don't allow customising headers.

For now the most common option is to use Vercel's Serve, but it has its own limitations.
